hal/network/ncp_client/sara/sara_ncp_client.cpp

```diff
@@ -2079,6 +2060,21 @@ int SaraNcpClient::enterDataMode() {
     return r;
 }
 
+int SaraNcpClient::dataModeError(int error) {
+    if (ncpId() == PLATFORM_NCP_SARA_R410 && error == SYSTEM_ERROR_PPP_NO_CARRIER_IN_NETWORK_PHASE) {
+        CHECK_TRUE(connectionState() == NcpConnectionState::CONNECTED, SYSTEM_ERROR_INVALID_STATE);
+        // FIXME: this is a workaround for some R410 firmware versions where the PPP session suddenly dies
+        // in network phase after the first IPCP ConfReq. For some reason CGATT=0/1 helps.
+        const NcpClientLock lock(this);
+        CHECK_FALSE(cgattWorkaroundApplied_, SYSTEM_ERROR_INVALID_STATE);
+        CHECK_PARSER_OK(parser_.execCommand(UBLOX_CFUN_TIMEOUT, "AT+CGATT=0"));
+        HAL_Delay_Milliseconds(1000);
+        CHECK_PARSER_OK(parser_.execCommand(UBLOX_CFUN_TIMEOUT, "AT+CGATT=1"));
+        cgattWorkaroundApplied_ = true;
```

Comment:
This line cgattWorkaroundApplied_ = true; may never get run if for any reason, the `CHECK_PARSER_OK` checks error out for `CGATT=0` or for `CGATT=1`. Should we perhaps account for any unforeseen circumstances where only CGATT=0 returns OK but `AT+CGATT=1` errors out?

Reply:
That's a valid concern, but because of `CHECK_PARSER_OK` we'll error out completely, reset the modem etc.

Reply:
And then it will get reset anyway in `connectionState()`.

Reply:
> That's a valid concern, but because of CHECK_PARSER_OK we'll error out completely, reset the modem etc.
Is it correct to say that `celMan_->ncpClient()->dataModeError(pppErrorToSystem(data));` is not actually getting the return error value from `dataModeError` in case `CHECK_PARSER_OK` fails?

Reply:
It doesn't care much about it, yes. We will set `parserError()` though if that `CHECK_PARSER_OK` fails, hmm. But that won't cause a modem reset if `checkParser()` gets a response from the modem. Any suggestions? I think if we fail to `CGATT=1`, we'll just repeat a few times (if we are still registered) and this should recover anyway

Reply:
That's my understanding as well. Clearing the `cgattWorkaroundApplied_ to true` at the end of the function makes sense to me so I don't think we need to change that. How about the following edge-case scenario:
- We run CGATT=0. At this very moment, _separately_, network coverage is lost so we won't be able to register and obviously CGATT=1 will fail. Need to confirm the ncp client smoothly handles this (if you think this is a valid edge case scenario)

Reply:
@keeramis How about we do `CGATT=0` here but instead set `CGATT=1` in `enterDataMode` before dialing if it's 0? Can you check that it's normally 1 if we don't interfere with anything?

Reply:
Default state is `1`
```
0000033524 [net.pppncp] TRACE: State changed event: 2
0000033528 [ncp.at] TRACE: > AT+CIMI
0000033528 [net.ppp.client] TRACE: PPP thread event LOWER_UP data=0
0000033539 [net.ppp.client] TRACE: State READY -> CONNECT
0000033544 [net.ppp.client] TRACE: State CONNECT -> CONNECTING
0000033536 [ncp.at] TRACE: < 310410232183811
0000033552 [ncp.at] TRACE: < OK
0000033556 [ncp.at] TRACE: > AT
0000033563 [ncp.at] TRACE: < OK
0000034565 [ncp.at] TRACE: > AT
0000034572 [ncp.at] TRACE: < OK
0000034573 [ncp.at] TRACE: > AT+CGATT?
0000034581 [ncp.at] TRACE: < +CGATT: 1
0000034582 [ncp.at] TRACE: < OK
0000034583 [ncp.at] TRACE: > ATD*99***1#
0000034589 [ncp.at] TRACE: < CONNECT 150000000
0000034594 [net.ppp.client] TRACE: PPP phase -> Initialize
0000034596 [net.ppp.client] TRACE: PPP phase -> Establish
0000034601 [lwip.ppp] TRACE: sent [LCP ConfReq id=0x1 <asyncmap 0x0> <magic 0x58527b14> <pcomp> <accomp>]

```

Reply:
Mind making the change while I look into the tracker flash overflow?

Reply:
This is the log from when I was able to test it out from a previous commit. I don't see any objections to set it in `enterDataMode()` because we anyway enter it. I don't know if the timing could affect it compared to what you have now, but I guess adding it before ATD is fine.. just gives it a little extra time between `CGATT=0` and `CGATT=1` than the hardcoded 1 sec delay.. 
```
0000080097 [hal] TRACE: sent [IPCP ConfReq id=0x1 <addr 0.0.0.0> <msdns1 8.8.8.8> <msdns2 8.8.4.4>]
0000080114 [net.ppp.client] ERROR: NO CARRIER in network PPP phase
0000080115 [net.ppp.client] TRACE: PPP phase -> Terminate
0000080121 [net.ppp.client] TRACE: PPP phase -> Establish
0000080125 [hal] TRACE: sent [LCP TermReq id=0x2 "User request"]
0000080131 [net.ppp.client] TRACE: PPP thread event ERROR data=100
0000080138 [net.pppncp] ERROR: PPP error event data=100
0000080142 [ncp.at] TRACE: > AT+CGATT=0
0000080235 [ncp.at] TRACE: < OK
0000081236 [ncp.at] TRACE: > AT+CGATT=1
0000081237 [ncp.at] TRACE: < +CEREG: 2
0000082160 [ncp.at] TRACE: < OK
0000082161 [ncp.at] TRACE: < +CEREG: 1,"6A33","7F0B50F",7
0000082164 [ncp.at] TRACE: > AT+CIMI
0000082169 [ncp.at] TRACE: < 310410232183811
0000082172 [ncp.at] TRACE: < OK
0000086131 [hal] TRACE: sent [LCP TermReq id=0x3 "User request"]
0000092133 [net.ppp.client] TRACE: PPP phase -> Disconnect
0000092134 [net.ppp.client] TRACE: PPP phase -> Dead
0000092141 [net.ppp.client] TRACE: PPP thread event ERROR data=5
0000092145 [net.pppncp] ERROR: PPP error event data=5
0000092150 [net.ppp.client] TRACE: PPP thread event DOWN data=0
0000092156 [net.ppp.client] TRACE: State CONNECTING -> DISCONNECT
0000092161 [net.ppp.client] TRACE: State DISCONNECT -> DISCONNECTING
0000092167 [net.ppp.client] TRACE: PPP thread event ERROR data=5
0000092173 [net.pppncp] ERROR: PPP error event data=5
0000092178 [net.ppp.client] TRACE: PPP thread event DOWN data=0
0000092184 [net.ppp.client] TRACE: State DISCONNECTING -> DISCONNECTED
0000092189 [net.ppp.client] TRACE: State DISCONNECTED -> CONNECT
0000092294 [net.ppp.client] TRACE: State CONNECT -> CONNECTING
0000092296 [ncp.at] TRACE: > AT
0000092303 [ncp.at] TRACE: < OK
0000093304 [ncp.at] TRACE: > AT
0000093311 [ncp.at] TRACE: < OK
0000093312 [ncp.at] TRACE: > ATD*99***1#
0000093322 [ncp.at] TRACE: < CONNECT 150000000
0000093322 [net.ppp.client] TRACE: PPP phase -> Initialize
0000093326 [net.ppp.client] TRACE: PPP phase -> Establish
```

Reply:
Let's do both? :) Keep the current logic CGATT=0/CGATT=1, but double check in `enterDataMode()` just in case.

Reply:
10000% that's what I was about to say :) I will make the change.

Reply:
@avtolstoy @scott-brust - Added the change here https://github.com/particle-iot/device-os/commit/c15e4f18b45ec1a8778c6978f923cd18849791df

Please note that IF `CGATT is not 1` and when we set it to `1`, it takes a moment or two for the modem to settle in as you see a bunch of registrations and de-registrations immediately after this command is set. So I added a 1sec delay and also set it as soon as we enter the `enterDataMode()`. I am not sure if we can reliably check that we are "stable" after CGATT=1. For testing though, I hardcoded CGATT=0 in enterDataMode() and this sequence works as follows:

```
0000055765 [net.ppp.client] TRACE: PPP thread event LOWER_UP data=0
0000055781 [net.ppp.client] TRACE: State READY -> CONNECT
0000055786 [net.ppp.client] TRACE: State CONNECT -> CONNECTING
0000055792 [ncp.client] TRACE: Entered enterDataMode()
0000055796 [ncp.at] TRACE: > AT
0000055801 [ncp.at] TRACE: < OK
0000055802 [ncp.at] TRACE: > AT+CGATT=0
0000055896 [ncp.at] TRACE: < OK
0000055897 [ncp.at] TRACE: > AT+CGATT?
0000055897 [ncp.at] TRACE: < +CEREG: 2
0000055905 [ncp.at] TRACE: < +CGATT: 0
0000055905 [ncp.at] TRACE: < OK
0000055910 [ncp.at] TRACE: > AT+CGATT=1
0000056780 [ncp.at] TRACE: < OK
0000058780 [ncp.at] TRACE: > AT
0000058787 [ncp.at] TRACE: < OK
0000058788 [ncp.at] TRACE: > ATD*99***1#
0000058797 [ncp.at] TRACE: < CONNECT 150000000
0000058798 [net.ppp.client] TRACE: PPP phase -> Initialize
0000058803 [net.ppp.client] TRACE: PPP phase -> Establish
0000058808 [lwip.ppp] TRACE: sent [LCP ConfReq id=0x1 <asyncmap 0x0> <magic 0x5852d9a0> <pcomp> <accomp>]
0000058799 [ncp.at] TRACE: < +CEREG: 1,"6A33","7F0B50F",7
0000058824 [ncp.at] TRACE: > AT+CIMI
0000058829 [ncp.at] TRACE: < 310410232183811
0000058830 [ncp.at] TRACE: < OK
0000058978 [lwip.ppp] TRACE: rcvd [LCP ConfReq id=0x2 <asyncmap 0x0> <auth chap MD5> <magic 0x9d68cc8b> <pcomp> <accomp>]
0000058986 [lwip.ppp] TRACE: sent [LCP ConfAck id=0x2 <asyncmap 0x0> <auth chap MD5> <magic 0x9d68cc8b> <pcomp> <accomp>]
0000058997 [lwip.ppp] TRACE: rcvd [LCP ConfAck id=0x1 <asyncmap 0x0> <magic 0x5852d9a0> <pcomp> <accomp>]
0000059008 [lwip.ppp] TRACE: sent [LCP EchoReq id=0x0 magic=0x5852d9a0]
0000059013 [net.ppp.client] TRACE: PPP phase -> Authenticate
0000059299 [lwip.ppp] TRACE: rcvd [LCP DiscReq id=0x3 magic=0x9d68cc8b]
0000059301 [lwip.ppp] TRACE: rcvd [CHAP Challenge id=0x1 <bc097177f7fc596bffd51450af4df162>, name = "UMTS_CHAP_SRVR"]
0000059312 [lwip.ppp] TRACE: sent [CHAP Response id=0x1 <4245a2caf15a9f00ccdee334bfa0c15d>, name = ""]
0000059323 [lwip.ppp] TRACE: rcvd [LCP EchoRep id=0x0 magic=0x9d68cc8b 58 52 d9 a0]
0000059618 [lwip.ppp] TRACE: rcvd [CHAP Success id=0x1 ""]
0000059619 [net.ppp.client] TRACE: PPP phase -> Network
0000059625 [lwip.ppp] TRACE: sent [IPCP ConfReq id=0x1 <addr 0.0.0.0> <msdns1 8.8.8.8> <msdns2 8.8.4.4>]
0000059938 [lwip.ppp] TRACE: rcvd [IPCP ConfReq id=0x1]
0000059940 [lwip.ppp] TRACE: sent [IPCP ConfAck id=0x1]
0000059944 [lwip.ppp] TRACE: rcvd [IPCP ConfNak id=0x1 <addr 10.217.113.16> <msdns1 174.47.40.106> <msdns2 174.47.20.106>]
0000059955 [lwip.ppp] TRACE: sent [IPCP ConfReq id=0x2 <addr 10.217.113.16> <msdns1 174.47.40.106> <msdns2 174.47.20.106>]
0000060258 [lwip.ppp] TRACE: rcvd [IPCP ConfAck id=0x2 <addr 10.217.113.16> <msdns1 174.47.40.106> <msdns2 174.47.20.106>]
0000060266 [system.nm] INFO: State changed: IFACE_UP -> IFACE_LINK_UP
0000060273 [net.ppp.client] TRACE: PPP thread event UP data=0
0000060277 [net.ppp.client] TRACE: State CONNECTING -> CONNECTED
0000060283 [net.pppncp] TRACE: Negotiated MTU: 1500
0000060273 [system.nm] INFO: State changed: IFACE_LINK_UP -> IP_CONFIGURED
0000060294 [net.ppp.client] TRACE: PPP phase -> Running

```